docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -95,51 +96,94 @@ We will know we’ve succeeded when we can benchmark speed increases and success
 
-##### FR1
+Example of the enviornment variable being turned on:
 
-##### FR2
+```
+export AZURE_OS_CACHING=true
+```
 
-#### Non-Functional Requirements
+The controller will maintain a timestamp in each AzureMachinePool and AzureMachineTemplate, and when the current time is the chosen interval ahead or more, the controller will perform the caching. Since the current controller manager requeues all objects every ten minutes by default the objects will be requeued shortly after its due time to be recached. This is because typically we expect to cache every 24 hours and it is very unexpected that this won't be frequent enough considering normal patch rates. When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image. After, it will store the current time as its timestamp.
 
-##### NFR1
+Example of how the timestamp will be maintained in the AzureMachinePool and AzureMachineTemplates:
 
-##### NFR2
+```yaml
+status:
+  conditions:
+  - lastTransitionTime: "2023-06-12T23:14:55Z"
+    status: "True"
+    type: LastOsCache
+```
 
-### Implementation Details/Notes/Constraints
+As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachineTemplates.
 
-The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachine Controller and AzureMachinePool Controller.
+In terms of how a healthy node would be selected, there is already a timestamp on each AzureMachinePool from its status.conditions.lastTransitionTime type ScaleSetModelUpdated as to when they’ve last been patched. There is also a timestamp on its status.conditions.lastTransitionTime type ScaleSetRunning as to when it last started running. For AzureMachine instances, we would need to add these timestamps or track it programmatically, below I have put an example of the yaml changes for storing the timestamp. An ideal node would be one which has been patched since the last prototype went into service and is running and healthy. Whichever node has been running and healthy for the longest amount of time since the last patch and has the patch applied should be chosen as it’s the most overall stable. As the prototype is always from a successfully healthy and working node the image is always known to be working before being chosen for replication.
```

Comment:
> There is also a timestamp on its status.conditions.lastTransitionTime type ScaleSetRunning as to when it last started running.

This is another condition on the entire AzureMachinePool where we would need how long each individual VM has been running, since not all VMs are expected to start running at the same time.

Reply:
Fixed this whole area, met with you and Jack Friday and got this all in a much better/more solid state by looking at the azuremachinepoolmachines themselves and finding this 'updated' data on each of them.